### ISSUE TITLE
nup feature?

### ISSUE THREAD

**Issue by colourful-land** [qpdf/qpdf#152]
Is it possible to use qpdf for the function provided by pdfjam or pdfnup?

These tools would lay multiple pages of content on a single pdf page. It is not for the purpose of merging multiple pdf files into a single pdf file. It usually only take a single pdf file as input, and a single pdf file as output which has less pages than the input.

Thanks.

**Comment by jberkenbilt** [comment]
The Linux Printing project's cups-filters includes code that uses the qpdf library to do this. The library definitely supports everything you would need to code this, but there's no command-line interface that does it. It could be added, but adding this is not a priority right now.

**Comment by colourful-land** [comment]
Thanks for the answer!

There are a few advantages implementing this, together with a few other functions offered by TexLive, for getting qpdf more often used by OS X users.

Many OS X users ended up installing that 4.3GB TexLive just to use the pdf utilities like crop, rotate and nup (pdfnup and pdfjam). These users would use qpdf based utilities if they are available.

Not to say qpdf should provide them; I'm just saying that there is a market for these:)

**Comment by jberkenbilt** [comment]
qpdf 7.0.0 will include a rotate option. It's in the current beta. I implemented by popular request, and actually, I used it myself just the other day and was glad to have it. :-) The next thing I'm going to work on is form flattening, and that activity will be the first thing in qpdf that touches content streams. Once that is established, the path to this type of feature will be shorter. I will definitely consider this as a possible enhancement. Optionally, feel free to suggest how you think it should work. No guarantees I'll take the suggestion, but I will definitely consider it as I think it through. Thanks.

**Comment by CsatiZoltan** [comment]
@jberkenbilt  Any update on this issue? I use `qpdf` version 10.6.3, which has the `--rotate` option. Nevertheless, I still cannot place multiple pages from the input PDF into a single page for the output PDF. Do you have a solution?

**Comment by jberkenbilt** [comment]
I'm inching closer to this. I'm wrapping up the last significant feature before it, then this is next. Realistically a few months away, but I think it will happen sometime after the qpdf 11 release.

**Comment by weiwu-zhang** [comment]
I welcome this feature, just to let you know there is one typical use case of n-up: to bind a booklet.

Let's say the input has 8 pages, the output will have 4 pages, but page 1 and 8 are collected into a big page, and page 2 and 7 are collected into another.

The idea is once all printing is done and double-sided, one can put a binder or stable at the centre of the page, and fold the paper into a 1/2 sized booklet, where the page numbers will be natural.

This typically requires less than 32 pages and if the user has 320 pages, it will be bound as a book. Book binding is a far less typical use-case but I write here just in case:
> pages are separated into 10 groups each of 32 pages, and bound in the centre 10 times, then glued together to form a book.

**Comment by jberkenbilt** [comment]
Yes, I'm aware of the book binding feature. I used to use psutils and know about the psbook command as well as the ability of pstops to arrange pages that way. I'll probably review the command-line syntax for those commands when I figure out how I want qpdf's CLI to work for this. Thanks for the reminder.

**Comment by AlisterH** [comment]
Note that paperjam (which uses libqpdf) also has this function.

**Comment by jberkenbilt** [comment]
This issue is part of the qpdf pages epic. If you are interested in following, please see #1104.
